File: MCMC Normal-lik-Norm-Gam-prior.R
x <- rnorm(15, mean = 5.7, sd = sqrt(16))
mean(x)
var(x)

init <- list()
init$mu <- 0.00
init$sig2 <- 1.00

prior <- list()
prior$m.0 <- 4.00
prior$sig2.mu <- 10.00
prior$a <- 1
prior$b <- 3

update.mu <- function(sig2, x, sig2.mu, m.0){
	n <- length(x)
	x.bar <- mean(x)

	sig2.mu.post <- 1/(n/sig2 + 1/sig2.mu)
	m.0.post <- sig2.mu.post * (sum(x)/sig2 + m.0/sig2.mu)

	rnorm(1, m.0.post, sig2.mu.post)
}

update.sig <- function(a, b, x, mu){
	n <- length(x)
	a.post <- n/2 + a
	sumsq <- sum((x - mu)^2)
	b.post <- sumsq/2 + b

	1/rgamma(1, a.post, b.post)
}

n_iter <- 50000
mu_out <- numeric(n_iter)
sig2_out <- numeric(n_iter)

mu_now <- init$mu
sig2_now <- init$sig2

for(i in 1:n_iter){
	mu_now <- update.mu(sig2_now, x, prior$sig2.mu, prior$m.0)
	sig2_now <- update.sig(prior$a, prior$b, x, mu_now)

	mu_out[i] <- mu_now
	sig2_out[i] <- sig2_now
}

post <- cbind(mu = mu_out, sig2 = sig2_out)
post.burnin <- post[5001:50000,]

apply(post.burnin, 2, mean)

library(coda)
plot(as.mcmc(post.burnin[,1:2]))

quantile(post.burnin[,1], c(.05, .95))
quantile(post.burnin[,2], c(.05, .95))

HPDinterval(as.mcmc(post.burnin[,1]), prob = .95)
HPDinterval(as.mcmc(post.burnin[,2]), prob = .95)

File: MCMC Normal-lik-Norm-Gam-prior.py
#Bayesian Inference for the parameters of a normal distribution, using a normal-gamma prior.
import math
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

x = np.random.normal(loc = 5.7, scale = 4, size = 100)
plt.hist(x)

#print(x.var())
#print(x.mean())

#Initial values
init = [0.0, 1.0]

#Prior hyperparameters
priorMu = [0.0, 100.0]
priorSig = [0.1, 0.1]

#Defining functions for MCMC

#Function to update the mean
def updateMu(sig, x, priorMu):
    n = len(x)
    sigMuPost = 1/(n/sig + 1/priorMu[1])
    mMuPost = sigMuPost * (sum(x)/sig + priorMu[0]/priorMu[1])
    
    return np.random.normal(loc = mMuPost, scale = math.sqrt(sigMuPost), size = 1)

#Function to update the variance
def updateSig(mu, x, priorSig):
    n = len(x)
    aPost = n/2 + priorSig[0]
    sumsq = sum((x - mu)**2)
    bPost = sumsq/2 + priorSig[1]
    
    return 1/np.random.gamma(shape = aPost, scale = 1/bPost, size = 1)

#Defining MCMC function
def MCMC(x, priorMu, priorSig, init, niter = 50000, burnin = 25000):
    post = np.zeros(shape = (niter, 2))
    for i in range(niter):
        if i == 0 :
            post[i,0] = updateMu(init[1], x, priorMu)
            post[i,1] = updateSig(init[0], x, priorSig)
        else :
            post[i,0] = updateMu(post[i-1,1], x, priorMu)
            post[i,1] = updateSig(post[i-1,0], x, priorSig)
        print(i)
    
    return pd.DataFrame(post[(burnin):,], columns = ["mean", "variance"])
      
#How do I indicate I got out of the function?    
df = MCMC(x, priorMu, priorSig, init)
df.head()
df.mean()
df.var()

plt.plot(df[["mean"]])
plt.plot(df[["variance"]])
